server/chat-plugins/the-studio.ts

```diff
@@ -0,0 +1,613 @@
+	randrec: 'recommendation',
+	randomrecommendation: 'recommendation',
+	async recommendation(target, room, user) {
+		if (!recommendations.saved.length) {
+			throw new Chat.ErrorMessage(`There are no recommendations saved.`);
+		}
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		this.runBroadcast();
+		if (!target) {
+			return this.sendReply(`|html|${await Recs.render(Recs.getRandomRecommendation())}`);
+		}
+		const matches: Recommendation[] = [];
+		if (target) {
+			target = target.slice(0, 300);
+			const args = target.split(',');
+			for (const rec of recommendations.saved) {
+				if (!args.every(x => rec.tags.map(toID).includes(toID(x)))) continue;
+				matches.push(rec);
+			}
+		}
+		if (!matches.length) {
+			throw new Chat.ErrorMessage(`No matches found.`);
+		}
+		const sample = Utils.shuffle(matches)[0];
+		this.sendReply(`|html|${await Recs.render(sample)}`);
+	},
+	recommendationhelp: [
+		`/recommendation [key1, key2, key3, ...] - Displays a random recommendation that matches all keys, if one exists.`,
+		`If no arguments are provided, a random recommendation is shown.`,
+	],
+
+	likerec: 'likerecommendation',
+	likerecommendation(target, room, user, connection) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		const [artist, title] = target.split('|').map(x => x.trim());
+		if (!(artist && title)) return this.parse(`/help likerecommendation`);
+		Recs.likeRecommendation(artist, title, user);
+		this.sendReply(`You liked '${title}' by ${artist}.`);
+	},
+	likerecommendationhelp: [
+		`/likerecommendation artist | title - Upvotes a recommendation for the provided artist and title.`,
+	],
+
+	viewrecs: 'viewrecommendations',
+	viewrecommendations(target, room, user) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		this.parse(`/j view-recommendations-${targetRoom!.roomid}`);
+	},
+	viewrecommendationshelp: [
+		`/viewrecommendations OR /viewrecs - View all recommended songs.`,
+	],
+
+	viewsuggestions: 'viewsuggestedrecommendations',
+	viewsuggestedrecs: 'viewsuggestedrecommendations',
+	viewsuggestedrecommendations(target, room, user) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		this.parse(`/j view-suggestedrecommendations-${targetRoom!.roomid}`);
+	},
+	viewsuggestedrecommendationshelp: [
+		`/viewsuggestedrecommendations OR /viewsuggestions - View all suggested recommended songs. Requires: % @ * # &`,
+	],
+};
+
+export const pages: PageTable = {
+	async recommendations(query, user, connection) {
+		const room = this.requireRoom();
+		if (!room.checkModjoin(user)) {
+			throw new Chat.ErrorMessage(`Access denied.`);
+		}
+		if (!user.inRooms.has(room.roomid)) throw new Chat.ErrorMessage(`You must be in ${room.title} to view this page.`);
```

Comment:
Why block them from viewing it if they aren't in the room? Perhaps minor, but no other plugin does this, and people may want to view it without being in it, and `checkModjoin` should handle the rest of it.

Reply:
because the commands still trigger even if the user isnt in the room (similarly for other plugins like quotes)